eject from react-scripts and update to react 18

kimberlythegeek opened the pull request:
my assumption is that what I've done is not an optimal solution and I don't consider it 'ready to merge', as there are many changes made here and I'm not completely confident I made the best decisions in implementation.

@julienw as far as the `async` `await` for `act()` calls, if I don't do that, the test will pass, but it throws a 'not wrapped in act' warning'

namely when calling `.jest.runOnlyPendingTimers()`. I tried many workarounds for this, but the only thing I found to resolve the issue was wrapping the call in `act` and using `await` `async`

julienw commented:
> @julienw as far as the async await for act() calls, if I don't do that, the test will pass, but it throws a 'not wrapped in act' warning'

From my limited testing, it seems to be the case for some tests (eg `App.test.tsx` seems to need it) but not in others (the test `Search/fetchRevisionByID.test.tsx` seems to work fine without the await/async act).

My rough guess is that in the ones that need it, the pending timers trigger promises, and these promises resolution callback end up out of the `act` call.

It's not clear to me what the best way forward is, let's say that keeping await/async in these cases makes sense (with comments), but that it would be good to not use it when we don't need it. The rationale here is that I believe it's good to avoid "magic incantations", that in the future we don't know anymore why they're needed and we end up adding them everywhere, and we're afraid of removing them.

kimberlythegeek commented on package.json:
```diff
@@ -55,67 +47,55 @@
+    "webpack": "^5.74.0",
+    "webpack-cli": "^4.10.0",
+    "webpack-dev-server": "^4.9.3",
+    "webpack-merge": "^5.8.0"
```
webpack-dev-server is for running the app locally, webpack-merge is used to merge configuration files

kimberlythegeek commented on public/index.html:
```diff
@@ -2,28 +2,17 @@
   <head>
     <meta charset="utf-8" />
-    <link rel="icon" href="%PUBLIC_URL%/favicon.svg" type="image/svg+xml" />
+    <link rel="icon" href="favicon.svg" type="image/svg+xml" />
```
`%PUBLIC_URL%` is no longer needed when building with webpack

kimberlythegeek commented on public/index.html:
```diff
@@ -2,28 +2,17 @@
       manifest.json provides metadata used when your web app is installed on a
       user's mobile device or desktop. See https://developers.google.com/web/fundamentals/web-app-manifest/
     -->
-    <link rel="manifest" href="%PUBLIC_URL%/manifest.json" />
```
generated file left over from `create-react-app`

julienw replied:
You can remove the comment above too.

kimberlythegeek commented on src/__tests__/App.test.tsx:
```diff
@@ -1,12 +1,14 @@
-  test('Should render search view on default route', () => {
+  test('Should render search view on default route', async () => {
     render(<App />);
+    await act(async () => void jest.runOnlyPendingTimers());
```
this is sometime required to avoid `an update to ___ was not wrapped in act()` warnings

kimberlythegeek commented on src/__tests__/App.test.tsx:
```diff
@@ -24,11 +26,9 @@ describe('App', () => {
     });
 
     await user.click(darkModeButton);
-    jest.runOnlyPendingTimers();
```
no longer needed

kimberlythegeek commented on src/__tests__/App.test.tsx:
```diff
@@ -24,11 +26,9 @@ describe('App', () => {
     expect(screen.queryByTestId('Brightness7Icon')).toBeInTheDocument();
 
     await user.click(darkModeButton);
-    jest.runOnlyPendingTimers();
```
same

kimberlythegeek commented on src/__tests__/Search/__snapshots__/SearchResultsList.test.tsx.snap:
```diff
@@ -380,12 +370,9 @@ exports[`SearchResultsList should match snapshot 1`] = `
                 </li>
-                <span
-                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
-                />
```
this was altered due to a change in the await/act/run timers implementation. I believe this is an animation that is no longer visible

kimberlythegeek commented on src/__tests__/utils/fileTransformer.js:
```diff
@@ -2,6 +2,8 @@ const path = require('path');
-    return `module.exports = ${JSON.stringify(path.basename(filename))};`;
+    return {
+      code: `module.exports = ${JSON.stringify(path.basename(filename))};`,
+    };
```
```
    ● Invalid return value:
      `process()` or/and `processAsync()` method of code transformer found at 
      "/home/kimberly/Workspace/perfcompare/src/__tests__/utils/fileTransformer.js" 
      should return an object or a Promise resolving to an object. The object 
      must have `code` property with a string of processed code.
      This error may be caused by a breaking change in Jest 28:
      https://jestjs.io/docs/upgrading-to-jest28#transformer
      Code Transformation Documentation:
      https://jestjs.io/docs/code-transformation
```

kimberlythegeek commented on src/components/Search/SearchDropdown.tsx:
```diff
@@ -21,7 +21,7 @@ function SearchDropdown(props: SearchDropdownProps) {
             value={repoMap[key]}
             key={repoMap[key]}
-            onClick={(e) => handleChangeDropdown(e)}
+            onClick={(e) => void handleChangeDropdown(e)}
```
`Promise-returning function provided to attribute where a void return was expected`

kimberlythegeek commented on src/components/Search/SearchViewInit.tsx:
```diff
@@ -1,16 +1,17 @@
   const { repository } = props;
   useEffect(() => {
-    dispatch(fetchRecentRevisions(repository));
+    void dispatch(fetchRecentRevisions(repository));
```
`Promises must be awaited, end with a call to .catch, end with a call to .then with a rejection handler or be explicitly marked as ignored with the `void` operator.`

this promise is awaited in the thunk action

kimberlythegeek commented on src/hooks/app.ts:
```diff
@@ -0,0 +1,8 @@
+
+// Use throughout your app instead of plain `useDispatch` and `useSelector`
+export const useAppDispatch: () => AppDispatch = useDispatch;
+export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;
```
https://redux-toolkit.js.org/tutorials/typescript#define-typed-hooks

kimberlythegeek commented:
I realize this is a very large patch. It may be possible to break this down further, but I was unsure how to do so without breaking the build or the tests. I tried to document each notable change I made and the reason for the change.

julienw commented on package.json:
```diff
@@ -55,67 +47,55 @@
-    "@babel/preset-env": "^7.16.11",
-    "@babel/preset-react": "^7.16.7",
-    "@babel/preset-typescript": "^7.16.7",
-    "@swc/core": "^1.2.196",
```
My understanding is that they were recommending to install @swc/core (see their doc: https://swc.rs/docs/usage/jest)
But if that wotks without the entry in package.json I'm fine with that too, so that's up to you.

julienw commented on package.json:
```diff
@@ -55,67 +47,55 @@
+    "@typescript-eslint/parser": "^5.31.0",
+    "babel-loader": "^8.2.5",
+    "babel-preset-jest": "^28.1.3",
+    "case-sensitive-paths-webpack-plugin": "^2.4.0",
```
sounds a good idea

julienw commented on jest.config.ts:
```diff
@@ -0,0 +1,52 @@
+  transformIgnorePatterns: [
+    '[/\\\\]node_modules[/\\\\].+\\.(js|jsx|mjs|cjs|ts|tsx)$',
+    '^.+\\.module\\.(css|sass|scss)$',
+  ],
```
Keeping this sounds good to me, we can look at removing this in a future PR if we see it's not needed.

julienw commented on .babelrc:
```diff
@@ -0,0 +1,17 @@
+        "regenerator": true
+      }
+    ]
+  ],
```
a few notes about that (all optional for this patch):
1. my understanding is that `@babel/preset-env` will add this plugin automatically
2. `true` seems to be the default for the `regenerator` option
3. That said I doubt this will even be used if you target latest versions of chrome and firefox. :)

julienw commented on .babelrc:
```diff
@@ -0,0 +1,17 @@
+  ],
+  "presets": [
+    "@babel/preset-env",
+    ["@babel/preset-react", { "runtime": "automatic" }],
```
sounds great to me!

julienw commented on .babelrc:
```diff
@@ -0,0 +1,17 @@
+    ]
+  ],
+  "presets": [
+    "@babel/preset-env",
```
This is how we configure preset-env in the profiler:
https://github.com/firefox-devtools/profiler/blob/a299ff00ea661579e96843d4c3d8cc7772133bc5/.babelrc#L4-L13

optional for this patch, it's better to look at this in another PR -- no high priority either.

kimberlythegeek replied:
#202

julienw commented on .babelrc:
```diff
@@ -0,0 +1,17 @@
+    "@babel/preset-env",
+    ["@babel/preset-react", { "runtime": "automatic" }],
+    "@babel/preset-typescript",
+    "babel-preset-jest"
```
This seems to be automatically added when using babel-jest (which is also automatic when running jest and babel-jest is installed), so could possibly remove this line. Also I'm a bit worried that this transform is always present even when not running tests (that said I have no idea what this does).

update: I see that `babel-jest` is not installed in this project, so if everything is running fine with this config, I suggest to handle the proposed change in a future PR.

julienw reviewed (approved):
Great work Kimberly!
I don't see big problems, only a few adjustments that could be made in small PRs.
I also gave the app a try locally, this looked good to me, so let's use the new React API!

So I'd say, let's land this and adjust later!

kimberlythegeek commented:
I added a couple of changes--I noticed the tests were taking a longer time to run again (~20 seconds). The culprit seemed to be the material-ui imports in `SearchView.tsx` that were pulling in unused modules.
See: https://mui.com/material-ui/guides/minimizing-bundle-size/#option-1

I tried a couple of other things, but the only significant difference I noticed was when I updated these imports.

also updated the readme instructions to indicate that `npm run start` is the command to run the development server.

This was changed originally to fix the heroku deployment. The issue was that heroku uses `npm run start` to build the app, and `start` used `react-scripts`. This package was moved to `devDependencies` due to a security vulnerability, and heroku will only install production dependencies when building.

see: https://github.com/mozilla/perfcompare/pull/81#issue-1205807378
